File: examples/java/src/main/java/com/coinbase/cdp/examples/enduser/SignEvmHash.java
package com.coinbase.cdp.examples.enduser;

import com.coinbase.cdp.CdpClient;
import com.coinbase.cdp.examples.utils.EnvLoader;
import com.coinbase.cdp.openapi.model.SignEvmHashWithEndUserAccountRequest;

/**
 * Example: Sign an EVM hash on behalf of an end user.
 *
 * <p>This example requires the end user to have an active delegation on their account that allows
 * the developer to sign on their behalf.
 *
 * <p>Usage: ./gradlew runSignEvmHash --args="&lt;userId&gt;"
 */
public class SignEvmHash {

  public static void main(String[] args) throws Exception {
    if (args.length < 1) {
      System.err.println("Usage: SignEvmHash <userId>");
      System.exit(1);
    }

    String userId = args[0];
    EnvLoader.load();

    try (CdpClient cdp = CdpClient.create()) {
      // Get the end user to find their EVM address
      var endUser = cdp.endUser().getEndUser(userId);

      if (endUser.getEvmAccountObjects().isEmpty()) {
        System.err.println("End user has no EVM accounts.");
        System.exit(1);
      }

      String address = endUser.getEvmAccountObjects().get(0).getAddress();
      System.out.println("Signing hash for user: " + userId);
      System.out.println("  Address: " + address);
      System.out.println();

      var response =
          cdp.endUser()
              .signEvmHash(
                  userId,
                  new SignEvmHashWithEndUserAccountRequest()
                      .hash(
                          "0x0000000000000000000000000000000000000000000000000000000000000001")
                      .address(address));

      System.out.println("Hash signed successfully!");
      System.out.println("  Signature: " + response.getSignature());
    }
  }
}

File: examples/python/end_user/sign_evm_hash.py
# Usage: uv run python end_user/sign_evm_hash.py <user_id>
# Note: This example requires the end user to have an active delegation on their account
# that allows the developer to sign on their behalf.

import asyncio
import sys

from cdp import CdpClient
from dotenv import load_dotenv

load_dotenv()


async def main():
    if len(sys.argv) < 2:
        print("Usage: uv run python end_user/sign_evm_hash.py <user_id>")
        sys.exit(1)

    user_id = sys.argv[1]
    hash_to_sign = "0x" + "ab" * 32

    async with CdpClient() as cdp:
        try:
            # Get the end user to retrieve their EVM account address.
            end_user = await cdp.end_user.get_end_user(user_id=user_id)
            address = end_user.evm_accounts[0].address

            # Sign an EVM hash using the client method.
            result = await cdp.end_user.sign_evm_hash(
                user_id=user_id,
                hash=hash_to_sign,
                address=address,
            )

            print("Signature (client method):", result.signature)

            # Alternatively, sign an EVM hash using the account method.
            # result = await end_user.sign_evm_hash(
            #     hash=hash_to_sign,
            #     address=address,
            # )

            # print("Signature (account method):", result.signature)

        except Exception as e:
            print(f"Error signing EVM hash: {e}")
            raise e


asyncio.run(main())
